plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java

```diff
@@ -307,6 +317,50 @@ public void testMissingColumnsOnInsert()
         assertQuery("SELECT * FROM test_col_insert", "SELECT 1, 'val1', 'val2'");
     }
 
+    @Test
+    public void testTopNPushdown()
+    {
+        assertThat(query("SELECT orderkey FROM orders ORDER BY orderkey LIMIT 10"))
+                .ordered()
+                .isNotFullyPushedDown(TopNNode.class, ExchangeNode.class, ExchangeNode.class, LimitNode.class);
```

Comment:
This passes the test. Not entirely sure whether it is still a useful test, though.

Reply:
I think we can include few more tests from BCT with modification here, like order by DESC, multiple different orders etc.

Reply:
How is this different from testing that TopN pushdown doesn't happen? (which is already tested in superclass if SUPPORTS_TOPN_PUSHDOWN is false). Or is the difference that the plan looks different (the retained nodes look same to me).

i.e. would it be better to add the new cases to superclass itself? https://github.com/trinodb/trino/blob/41ff64c3d4f065c754d21af3f1ccd91c351a1240/plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java#L626-L631

Reply:
This is the original plan:
```
Output[orderkey]
│   Layout: [orderkey:bigint]
└─ TopN[10 by (orderkey ASC NULLS LAST)]
│   Layout: [orderkey:bigint]
└─ LocalExchange[SINGLE] ()
│   Layout: [orderkey:bigint]
└─ RemoteExchange[GATHER]
│   Layout: [orderkey:bigint]
└─ TopNPartial[10 by (orderkey ASC NULLS LAST)]
│   Layout: [orderkey:bigint]
└─ TableScan[phoenix:tpch.orders TPCH.ORDERS sortOrder=[ORDERKEY:bigint:BIGINT ASC NULLS LAST] limit=10 columns=[ORDERKEY:bigint:BIGINT]]
Layout: [orderkey:bigint]
orderkey := ORDERKEY:bigint:BIGINT
```

So this indeed verifies that the `topNPartial` was replaced by a `limit`

Reply:
Thanks @lhofhansl . Missed that the partial got removed.

Reply:
I wasn't sure either. Had to run the test without the change. :)

Reply:
or maybe we use `BasePlanTest` see https://github.com/trinodb/trino/blob/f221c030548d46a0a42ce6d4586d5a776126487b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestHivePlans.java#L139-L146 for example

Reply:
@lhofhansl Using `isNotFullyPushedDown` for asserting plan shape when it has nothing to do with pushdown seems a bit dirty.

Using something similar to what Piotr shows above is more explicit in the intent of the test.

Reply:
I think we can use the `isNotFullyPushedDown(PlanMatchPattern retainedSubplan)` method instead to provide the subplan tree instead of the list of classes

Reply:
Sorry, didn't see that comment.

Looked at that. Wouldn't I create the `PlanMatchPattern` pretty much the same as `isNotFullyPushedDown(Class<? extends PlanNode>... retainedNodes)` does internally?

Reply:
ya, I was just trying to see if we can get the test to resemble typical assertPlan tests.
We can extract these in a separate class which derives from BasePlanTest similar to TestHivePlans and verifies the plans using assertDistributedPlan.
But i'm wondering if it's ok to just add below method in QueryAssertions and use that to assert plan
```
public final QueryAssert matches(PlanMatchPattern expectedPlan)
        {
            transaction(runner.getTransactionManager(), runner.getAccessControl())
                    .execute(session, session -> {
                        Plan plan = runner.createPlan(session, query, WarningCollector.NOOP);
                        assertPlan(
                                session,
                                runner.getMetadata(),
                                noopStatsCalculator(),
                                plan,
                                expectedPlan);
                    });
            return this;
        }
```

Reply:
That would work.

Not sure what we are trying to test. The basic plan rewrite is done in #6634, here I just want to make that Phoenix is triggering that query rewrite. Assuming changes to #6634 would add tests, I would add the most generic version to the connectors, just making sure that the replacement actually happens.

As usually, I do not feel strongly :)  Happy to add a more explicit test as you suggest.

Reply:
#6634 did add tests with a mock connector, hive and later on tpch connector
Here we want to test that after partial pushdown of topN into table scan takes place, we detect that through the sorting ordering in the JDBC table handle and expose the right ConnectorTableProperties for it. This part is unique to this connector. We can assume that if the right properties are set, then the planner will do the expected query rewrite. The plan assertions are an indirect but convenient way to test that the right sorting properties were set.
PlanMatchPattern would allow us to explicitly verify the `preSortedInputs` in LimitNode.

Reply:
This is the only way I can get this to work:
```
        assertThat(query("SELECT orderkey FROM orders ORDER BY orderkey LIMIT 10"))
                .ordered()
                .matches(
                        output(
                                node(TopNNode.class,
                                        anyTree(
                                                node(LimitNode.class,
                                                        node(TableScanNode.class))))));
```
Not sure that is much better.

Reply:
We can make that slightly better by building more specific plan node patterns rather than just using Node.class, e.g. see the plan patterns used in assertions in `TestPartialTopNWithPresortedInput#testWithSortedTable`.
However, to me this is nice to have, not a blocker to merging the PR.

Reply:
Thanks! I had tried the more specific pattern, but in the time I had I couldn't get the plan to match. `TestPartialTopNWithPresortedInput` is a great example, though. I'll try that.